js/executor.js

```diff
@@ -1,355 +1,207 @@
-      await wasm.compile(
-        module,
-        {},
-      );
+        const cachedContract = localStorage.getItem(contractId);
+        const cachedInteractions = localStorage.getItem(`${contractId}-interactions`);
+
+        let [contract, interactions] = await Promise.all([
+            cachedContract ? JSON.parse(cachedContract) : this.loadContract(contractId, gateway),
+            cachedInteractions
+                ? JSON.parse(cachedInteractions)
+                : this.loadInteractions(contractId, height, gateway),
+        ]);
+
+        let updatePromise = [];
+        if (cachedInteractions) {
+            // So now we have the cached interactions
+            // but we still need to ensure that the cached interactions are up to date.
+            const lastEdge = interactions[interactions.length - 1];
+            if (lastEdge) {
+                updatePromise = this.updateInteractions(contractId, height, lastEdge.cursor, gateway);
+            }
+        }
 
-      let currState = encode(state);
-      for (const interaction of interactions) {
-        const input = interaction.node.tags.find((data) =>
-          data.name === "Input"
-        );
-        currState = wasm.call(
-          currState,
-          encode({
-            input,
-            caller: interaction.node.owner.address,
-          }),
-        );
-      }
+        if (!cachedContract) {
+            localStorage.setItem(contractId, JSON.stringify(contract));
+        }
+        if (!cachedInteractions) {
+            localStorage.setItem(
+                `${contractId}-interactions`,
+                JSON.stringify(interactions),
+            );
+        }
 
-      return currState;
-    case "application/octet-stream":
-      // TODO(perf): Streaming initalization
-      const res = await fetch(
-        "https://github.com/three-em/3em/raw/js_library/js/evm/evm.wasm",
-      );
-      const evmModule = new Uint8Array(await res.arrayBuffer());
-      const bytecode = hex(source);
-      const _storage = hex(state);
-      for (const interaction of interactions) {
-        const input = interaction.node.tags.find((data) =>
-          data.name === "Input"
-        );
+        const { source, state, type, tx } = contract;
+        let result = undefined;
+        switch (type) {
+            case "application/javascript":
+                const rt = new Runtime(source, state, {}, this, gateway);
+
+                // Slower than `rt.executeInteractions` but more readable
+                // 100 interactions in ~30.06ms.
+                //
+                // for (const interaction of interactions) {
+                //    const input = interaction.node.tags.find(data => data.name === "Input");
+                //    await rt.execute({ input, caller: interaction.node.owner.address });
+                // }
+
+                // Faster. At 100 interactions in about 3.68ms.
+                await rt.executeInteractions(interactions, tx);
```

Comment:
can we make sure `interactions` and `updatedInteractions` ABSOLUBTELY respect `height`, even if there's higher heights cached, we should have a filter `<= height`

Reply:
@littledivy